Mathlib/Data/Matroid/Map.lean

```diff
@@ -0,0 +1,670 @@
+  inferInstanceAs (M.map f f.injective.injOn).RkPos
+
+end mapEmbedding
+
+section mapEquiv
+
+variable {f : α ≃ β}
+
+/-- Map `M : Matroid α` across an equivalence `α ≃ β` -/
+def mapEquiv (M : Matroid α) (f : α ≃ β) : Matroid β := M.mapEmbedding f.toEmbedding
+
+@[simp] lemma mapEquiv_ground_eq (M : Matroid α) (f : α ≃ β) :
+    (M.mapEquiv f).E = f '' M.E := rfl
+
+lemma mapEquiv_eq_map (f : α ≃ β) : M.mapEquiv f = M.map f f.injective.injOn := rfl
+
+@[simp] lemma mapEquiv_indep_iff {I : Set β} : (M.mapEquiv f).Indep I ↔ M.Indep (f.symm '' I) := by
+  rw [mapEquiv_eq_map, map_indep_iff]
+  exact ⟨by rintro ⟨I, hI, rfl⟩; simpa, fun h ↦ ⟨_, h, by simp⟩⟩
+
+@[simp] lemma mapEquiv_dep_iff {D : Set β} : (M.mapEquiv f).Dep D ↔ M.Dep (f.symm '' D) := by
+  rw [mapEquiv_eq_map, map_dep_iff]
+  exact ⟨by rintro ⟨I, hI, rfl⟩; simpa, fun h ↦ ⟨_, h, by simp⟩⟩
+
+@[simp] lemma mapEquiv_base_iff {B : Set β} : (M.mapEquiv f).Base B ↔ M.Base (f.symm '' B) := by
+  rw [mapEquiv_eq_map, map_base_iff]
+  exact ⟨by rintro ⟨I, hI, rfl⟩; simpa, fun h ↦ ⟨_, h, by simp⟩⟩
+
+instance [M.Nonempty] {f : α ≃ β} : (M.mapEquiv f).Nonempty :=
+  inferInstanceAs (M.map f f.injective.injOn).Nonempty
+
+instance [M.Finite] {f : α ≃ β} : (M.mapEquiv f).Finite :=
+  inferInstanceAs (M.map f f.injective.injOn).Finite
+
+instance [M.Finitary] {f : α ≃ β} : (M.mapEquiv f).Finitary :=
+  inferInstanceAs (M.map f f.injective.injOn).Finitary
+
+instance [M.FiniteRk] {f : α ≃ β} : (M.mapEquiv f).FiniteRk :=
+  inferInstanceAs (M.map f f.injective.injOn).FiniteRk
+
+instance [M.RkPos] {f : α ≃ β} : (M.mapEquiv f).RkPos :=
+  inferInstanceAs (M.map f f.injective.injOn).RkPos
+
+end mapEquiv
+
+section restrictSubtype
+
+variable {E X I : Set α} {M N : Matroid α}
+
+/-- Given `M : Matroid α` and `X : Set α`, the restriction of `M` to `X`,
+viewed as a matroid on type `X` with ground set `univ`.
+Always isomorphic to `M ↾ X`. If `X = M.E`, then isomorphic to `M`. -/
+def restrictSubtype (M : Matroid α) (X : Set α) : Matroid X := (M ↾ X).comap (↑)
+
+@[simp] lemma restrictSubtype_ground : (M.restrictSubtype X).E = univ := by
+  simp [restrictSubtype]
+
+@[simp] lemma restrictSubtype_indep_iff {I : Set X} :
+    (M.restrictSubtype X).Indep I ↔ M.Indep ((↑) '' I) := by
+  simp [restrictSubtype, Subtype.val_injective.injOn]
+
+lemma restrictSubtype_indep_iff_of_subset (hIX : I ⊆ X) :
+    (M.restrictSubtype X).Indep (X ↓∩ I) ↔ M.Indep I := by
+  rw [restrictSubtype_indep_iff, image_preimage_eq_iff.2]; simpa
+
+lemma restrictSubtype_inter_indep_iff :
+    (M.restrictSubtype X).Indep (X ↓∩ I) ↔ M.Indep (X ∩ I) := by
+  simp [restrictSubtype, Subtype.val_injective.injOn]
+
+lemma eq_of_restrictSubtype_eq (hM : M.E = E) (hN : N.E = E)
+    (h : M.restrictSubtype E = N.restrictSubtype E) : M = N := by
+  subst hM
+  refine eq_of_indep_iff_indep_forall (by rw [hN]) (fun I hI ↦ ?_)
+  rwa [← restrictSubtype_indep_iff_of_subset hI, h, restrictSubtype_indep_iff_of_subset]
```

Comment:
If the matroid `N` is only used for this lemma in section `restrictSubtype`, wouldn't it be better to declare the variable here?